src/misc/reachability.c

```diff
@@ -0,0 +1,122 @@
+/* -*- mode: C -*-  */
+/* vim:set ts=4 sw=4 sts=4 et: */
+/*
+   IGraph library.
+   Copyright (C) 2024  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_adjlist.h"
+#include "igraph_community.h"
+#include "igraph_components.h"
+#include "igraph_interface.h"
+#include "igraph_reachability.h"
+
+void _print_vector_int(const igraph_vector_int_t *v) {
+    igraph_integer_t i, n = igraph_vector_int_size(v);
+    printf("(");
+    for (i=0; i < n; i++) {
+        printf(" %" IGRAPH_PRId, VECTOR(*v)[i]);
+    }
+    printf(" )\n");
+}
+
+igraph_integer_t igraph_bitset_popcount(igraph_vector_int_t *bitset)
+{
+    igraph_integer_t i, n, count = 0;
+    n = igraph_vector_int_size(bitset);
+    for (i = 0; i < n; ++i)
+    {
+        // TODO: Check if __cpuid claims the operation is supported by CPU
+        // https://learn.microsoft.com/en-us/cpp/intrinsics/popcnt16-popcnt-popcnt64?view=msvc-170
+        #ifdef _MSC_VER
+        count += __popcnt64(VECTOR(*bitset)[i]);
+        #else
+        count += __builtin_popcountll(VECTOR(*bitset)[i]);
+        //printf("Current count at %ld: %ld\n", i, (igraph_integer_t)__builtin_popcountll(VECTOR(*bitset)[i]));
+        #endif
+    }
+    return count;
+}
+
+igraph_error_t igraph_reachability_directed(
+    const igraph_t *graph,
+    igraph_vector_int_t *membership,
+    igraph_vector_int_t *csize,
+    igraph_integer_t *no_of_components,
+    igraph_vector_int_list_t *reach)
+{
+    igraph_integer_t no_of_nodes;
+    igraph_integer_t i, j, k, n;
+    igraph_adjlist_t adjlist, dag;
+    igraph_vector_int_t *dag_neighbours, *neighbours, *from_bitset, *to_bitset;
+
+    no_of_nodes = igraph_vcount(graph);
+
+    igraph_connected_components(graph, membership, csize, no_of_components, IGRAPH_STRONG);
+
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &adjlist, IGRAPH_OUT, IGRAPH_LOOPS_ONCE, IGRAPH_MULTIPLE));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &adjlist);
+
+    IGRAPH_CHECK(igraph_vector_int_list_init(reach, *no_of_components));
```

Comment:
I think we should handle gracefully when the user is not interested in the number of components and provides `NULL` as `no_of_components`. Actually, we should probably accept `NULL` for all arguments except `graph` and allocate our own data structures if needed.

Reply:
Note that the typical convention in igraph is that incoming arguments are already initialized and are owned by the caller. We should not re-initialize `reach` here; we should call `igaph_vector_int_list_resize()` instead. I believe this may be the reason for the memory leaks in the unit tests.

Reply:
Yeah that was a leftover from an older version, forgot to change to resize. Memory leak fixed now.

Reply:
Note that without knowing membership, it's not particularly useful to get the reachability set. For optimization of the average case, I instead work with and return the reachability of each strongly connected component. So to obtain the value for a vertex you need both the reachability set of its component and the membership to determine the component. Perhaps this is not the API you would prefer, but it does save memory in the vast majority of cases.

Reply:
I agree that output parameters should almost always be optional, and this is now implemented. The one exception is `membership`, which is necessary to interpret the result, as Arnar said.